File: Advanced/code/problem_12.cpp
#include<bits/stdc++.h>

using namespace std;
int trap(vector < int > & height) {
  int n = height.size();
  int left = 0, right = n - 1;
  int res = 0;
  int maxLeft = 0, maxRight = 0;
  while (left <= right) {
    if (height[left] <= height[right]) {
      if (height[left] >= maxLeft) {
        maxLeft = height[left];
      } else {
        res += maxLeft - height[left];
      }
      left++;
    } else {
      if (height[right] >= maxRight) {
        maxRight = height[right];
      } else {
        res += maxRight - height[right];
      }
      right--;
    }
  }
  return res;
}

int main() {
  vector < int > arr;
  arr = {0,1,0,2,1,0,1,3,2,1,2,1};
  cout << "The water that can be trapped is " << trap(arr) << endl;
}

// Output: The water that can be trapped is 6

// Time Complexity: O(N) because we are using 2 pointer approach.

// Space Complexity: O(1) because we are not using anything extra.

File: Advanced/code/problem_12.java
// Trapping Rain Water
// Prashant9420
class Solution {
    public int trap(int[] height) {
        int area=0,fakeArea=0;
        int d=height[0],dind=0;
        for(int i=0;i<height.length;i++)
        {
            if(height[i]>=d)
            {
                d=height[i];
                dind=i;
                fakeArea=area;
            }
            else{
                if(i==height.length-1){
                    
                    d=height[i];
                    while(i!=dind){
                        if(height[i]>=d){
                            d=height[i];
                        }
                        else{
                            fakeArea+=d-height[i];
                        }
                        i--;
                    }
                    break;
                }
                area+=d-height[i];
            }
        }


        return fakeArea;
    }
}
